### ISSUE TITLE
[Android] Blocking exception with latest Xamarin.Forms 3.5 when leaving a page (worked on 3.4)

### ISSUE THREAD

**Issue by dirivero** [alexrainman/CarouselView#489]
**Description**
After updating to XF 3.5, an exception occurs every time a NavigationPage having at least a Carousel (within a ContentPage) as part of its navigation stack leaves the visual tree.

**Platform**
Android (I've tried with API 23 and 26)

**Steps to reproduce**
1. Open a modal content page A (with a carousel) inside a navigation page
2. Close the modal page A

Alternative flow:
1. Within a top level navigation page push a content page A containing a carousel
2. Replace the root page with a new page or navigation page

**Result**
```
System.NotSupportedException: Unable to activate instance of type Com.ViewPagerIndicator.CirclePageIndicator from native handle 0xff91625c (key_handle 0x379aca3). ---> System.MissingMethodException: No constructor found for Com.ViewPagerIndicator.CirclePageIndicator::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership) ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
  at Java.Lang.Error: Exception of type 'Java.Lang.Error' was thrown.
  at java.lang.Error: Java callstack:
  at at md5cb1a4bab49bfe17ad6c0290694c1d310.CirclePageIndicator.n_onSaveInstanceState(Native Method)
  at at md5cb1a4bab49bfe17ad6c0290694c1d310.CirclePageIndicator.onSaveInstanceState(CirclePageIndicator.java:85)
  at at android.view.View.dispatchSaveInstanceState(View.java:14669)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3093)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3099)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3099)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3099)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3099)
  at at android.view.ViewGroup.dispatchSaveInstanceState(ViewGroup.java:3099)
  at at android.view.View.saveHierarchyState(View.java:14652)
  at at android.support.v4.app.FragmentManagerImpl.saveFragmentViewState(FragmentManager.java:2859)
  at at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1500)
  at at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1750)
  at at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:792)
  at at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2590)
  at at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2377)
  at at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2332)
  at at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2239)
  at at android.support.v4.app.FragmentManagerImpl.executePendingTransactions(FragmentManager.java:781)
  at at mono.java.lang.RunnableImplementor.n_run(Native Method)
  at at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:30)
  at at android.os.Handler.handleCallback(Handler.java:739)
  at at android.os.Handler.dispatchMessage(Handler.java:95)
  at at android.os.Looper.loop(Looper.java:148)
  at at android.app.ActivityThread.main(ActivityThread.java:5417)
  at at java.lang.reflect.Method.invoke(Native Method)
  at at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
  at at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
  --- End of inner exception stack trace ---
  at Java.Interop.TypeManager.CreateProxy (System.Type type, System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer) [0x00055] in <ea97b6d670fe4c1ba9dadb89b2fa0103>:0
  at Java.Interop.TypeManager.CreateInstance (System.IntPtr handle, Android.Runtime.JniHandleOwnership transfer, System.Type targetType) [0x00116] in <ea97b6d670fe4c1ba9dadb89b2fa0103>:0
  --- End of inner exception stack trace ---
```

**Sample project**
I've created a sample minimal project exercising both flows. You can find it here:
https://drive.google.com/open?id=14TiTnEuucfs28E2lKd0SJAbHZ_V397uF

I'm guessing that that the save/restore override in this class is causing the problem: https://github.com/alexrainman/CirclePageIndicator/blob/master/Com.ViewPagerIndicator/CirclePageIndicator.cs

Thanks for this great component!

**Comment by quentinR0bert** [comment]
The same but he respond on this sub-project (https://github.com/alexrainman/CirclePageIndicator/issues/1)
He will release soon a update i think

**Comment by lewcianci** [comment]
In case you are like me and you couldn't wait, you can use this instead: https://github.com/AndreiMisiukevich/CardView

I think it has some benefits.  It doesn't use an Android specific implementation for the carouselview, it achieves it with built in xamarin layouts.  So I think it would be more resilient to change.  I just downloaded it and changed my nuget packages over and everything came to life.

Plus you get a choice of more visual flair (card view or carousel view for instance).  I like it.

**Comment by quentinR0bert** [comment]
This issue is not for making Ads for CardView. Stop using each issue for publicity.

The two plugin make the same thing yes but it's not use the same way.

**Comment by lewcianci** [comment]
Relax @quentinR0bert , nobody is making "ads for cardview".  I just thought some people, like myself, would like to move to XF 3.5 and need to find functional dependencies.  I have no association with CardView, I'm just trying to make my stuff work with XF 3.5, thats all.  Its a workaround for the issue at hand, so its a valid comment on an issue like this.

**Comment by alexrainman** [comment]
Guys, i am very close to release a new version and it will be compatible with XF 3.5 and with lot of new functionality.

**Comment by NiknakSi** [comment]
Because I'm impatient I did a little digging and made a fix for this in a branch in my fork.

https://github.com/NiknakSi/CarouselView/commit/ecbc4c5f7e99183ed2c394aee6dcaaa1815f7f4f

Basically I just brought the CirclePageIndicator code into the CarouselView solution temporarily and added the missing ctor overload. If you're desperate for a fix then I think this does the trick - I've been testing with XF 3.5.0.129452 and haven't had a crash yet.

**Comment by Nico04** [comment]
I've got the same issue, I hope your patch will be released soon :)
Thanks again for your work !

**Comment by alexrainman** [comment]
I am almost ready for a new release but have been traveling. Will keep you guys posted.

**Comment by ryanherman** [comment]
@alexrainman any idea when we can expect the update?  I know your busy, even if its a few weeks so I know how to plan.  Thank you

**Comment by alexrainman** [comment]
I think by end of week.

**Comment by Maldorami** [comment]
@alexrainman The update has contemplated this issue? 
"Unhandled Exception:
Java.Lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first. occurred"

I have a gallery like Facebook, I just open a pic (create the carousel), 2 o 3 swaps later thow this error, if you need the source code just tell me.
(Working with Xam < 3.5 and VS2019)

**Comment by Nico04** [comment]
Any news on a release? As you said it was coming soon I postponed my client's release but we've reached the end of the the week without any news. 
Any idea of a release date please?

**Comment by sthewissen** [comment]
Also waiting for this, so would be curious to hear about a release @alexrainman :)

**Comment by NelsonSantos** [comment]
> Also waiting for this, so would be curious to hear about a release @alexrainman :)

Me too! :)

**Comment by so3da** [comment]
@alexrainman , when we can expect the hotfix, please?

**Comment by TohWei** [comment]
@alexrainman  Please provide the fix asap. My app is heavily rely on the carousel view. At the moment is working fine for iOS but not Android. Please. 

This is the error I got under Xamarin.Forms 3.5.0.169047: 
System.NotSupportedException:
"Unable to activate instance of type Com.ViewPagerIndicator.CirclePageIndicator from native handle 0x7fefec6694 (key_handle 0x3375ed5)."

**Comment by alexrainman** [comment]
Sorry guys, i have been sick. Will come back to the project later today and i expect to release by end of week.

**Comment by NelsonSantos** [comment]
Hi guys...

I've tryed update the CirclePageIndicator nuget version to 1.0.3 version from nuget and seens that, after make some tests, the problem doesn't occour.

There are somebody that could apply this procedure to see if works for you too?

**Comment by TohWei** [comment]
I just added the Xamarin.Android.CirclePageIndicator (Com.ViewPagerIndicator CirclePageIndicator for Xamarin.Android.). version 1.0.3.  

It has fixed my problem.  It is working now. Thanks.

**Comment by alexrainman** [comment]
I updated CirclePageIndicator nuget as part of my fix. Glad it fixed yours too.

**Comment by Nico04** [comment]
Thanks @TohWei , I've done the same as you and it also fix the bug.
Thanks all

**Comment by kerrpeter** [comment]
When I upgrade Xamarin.Android.CirclePageIndicator to 1.0.3 I get the following Build Error:

Can not resolve reference: `Xamarin.Android.Support.ViewPager`, referenced by `Com.ViewPagerIndicator`. Please add a NuGet package or assembly reference for `Xamarin.Android.Support.ViewPager`, or remove the reference to `Com.ViewPagerIndicator`.	

I never had Xamarin.Android.Support.ViewPager installed before. So I tried adding it to my Android project but I still get the same error. What am I doing wrong?

**Comment by NelsonSantos** [comment]
> When I upgrade Xamarin.Android.CirclePageIndicator to 1.0.3 I get the following Build Error:
> 
> Can not resolve reference: `Xamarin.Android.Support.ViewPager`, referenced by `Com.ViewPagerIndicator`. Please add a NuGet package or assembly reference for `Xamarin.Android.Support.ViewPager`, or remove the reference to `Com.ViewPagerIndicator`.
> 
> I never had Xamarin.Android.Support.ViewPager installed before. So I tried adding it to my Android project but I still get the same error. What am I doing wrong?

Whats is your Xamarin.Android.Support.XXX dependencies version? 

All of that files in my project are based on 28.0.0.1. 

See if an update solve your problem.

**Comment by NelsonSantos** [comment]
Look at CirclePageIndicator dependencies needs a Xamarin.Android.Support.v4 at >= 28.0.0.1.

![image](https://user-images.githubusercontent.com/3474662/53909420-69d9cc80-4030-11e9-81a3-afc340b3e5f0.png)

**Comment by vjsystematix** [comment]
@NelsonSantos 

Are you find any solution on the same as i am also getting the same error after installing Xamarin.Android.CirclePageIndicator.

**Comment by NelsonSantos** [comment]
@vjsystematix Did you have tryed update Xamarin Support files to the version 28.0.0.1?

Here for me nothing occours. Works like a charm!

**Comment by vjsystematix** [comment]
Yes, I have updated all the files still facing the same issue. Please suggest

**Comment by kerrpeter** [comment]
Yeah installing Xamarin.Android.CirclePageIndicator 1.0.3 automatically installed those dependencies of Xamarin Android Support version 28.0.0.1
It did not automatically install  Xamarin.Android.Support.ViewPager, so where is that dependency coming from?

**Comment by kannamraju** [comment]
I am using .Net standard and both these packages (CirclePageIndicator and Xamarin.Android.Support.v4 ) are not compatible with  netstandard2.0

**Comment by kannamraju** [comment]
Can not resolve reference: `Xamarin.Android.Support.ViewPager`, referenced by `Com.ViewPagerIndicator`. Please add a NuGet package or assembly reference for `Xamarin.Android.Support.ViewPager`, or remove the reference to `Com.ViewPagerIndicator`.	CaptureForNurseries.Android

I am getting this error now even after installing `Xamarin.Android.Support.ViewPager`

Waiting for your release.

**Comment by NelsonSantos** [comment]
@vjsystematix @kerrpeter 

When I updated Xamarin.Forms to version 3.5 I'm already update all of Xamarin Android Support files to version 28.0.0.1, was when I run the project and noticed the problem of this thread.

So I've searched for solutions untill this of the update of nuget CirclePageIndicator that worked.

I think that this is the fact I belive it worked on my project. I have made this update of all Xamarin Support files before the update of nuget CirclePageIndicator.

Some times, I need to clean and do a delete of bin/obj folders of Android project to work. This has become a habit of mine!

**Comment by kerrpeter** [comment]
I got it working by changing my Target Framework to be Android 9.0 and re-installing the support packages using the following in Package manager:
Update-Package Xamarin.Android.Support.v4 -Reinstall

I also resinstalled the Xamarin.Android.Support.v7 packages I had installed. I ended up with some additional packages in project that were not there before, so maybe this was the probelm.
Hope it helps someone else.

**Comment by vikramupkarak** [comment]
@kerrpeter - Hi i did same process that you mention but still i getting the same error

**Comment by vikramupkarak** [comment]
> I updated CirclePageIndicator nuget as part of my fix. Glad it fixed yours too.

My issue haven't resolved after update this package . i am still getting below error.

Can not resolve reference: Xamarin.Android.Support.ViewPager, referenced by Com.ViewPagerIndicator. Please add a NuGet package or assembly reference for Xamarin.Android.Support.ViewPager, or remove the reference to Com.ViewPagerIndicator

**Comment by so3da** [comment]
Hi @alexrainman , Do we have any final solution for this issue, kindly we need your help?

**Comment by aimore** [comment]
@so3da add the CirclePageIndicator Nuget package version 1.0.3
it will fix the problem

**Comment by tunbeo** [comment]
Downgraded to XF 3.4 and everything is ok

**Comment by Ruddy2007** [comment]
@alexrainman Update on this?

**Comment by nrb1238** [comment]
@alexrainman 

Any Updates on this?

**Comment by CarLoOSX** [comment]
any pre-release ?

**Comment by LeoJHarris** [comment]
@so3da Fixed my issue as well: add the CirclePageIndicator Nuget package version 1.0.3

**Comment by LeoJHarris** [comment]
@aimore that does seem to resolve the issue not crashing when navigating away but navigating back the carousel doesn't have any content anymore.

**Comment by LeoJHarris** [comment]
@alexrainman also like to know when we can at least expect / hope to have something on this?

**Comment by PWaliaDev** [comment]
@alexrainman  This issue still not fixed. This is blocking. I am not able to update Xamarin.Forms version because of this. 

And if I add Xamarin.Android.Support.ViewPager then get below error.

Error XA2002: Can not resolve reference: `Xamarin.Android.Support.ViewPager`, referenced by `Com.ViewPagerIndicator`. Please add a NuGet package or assembly reference for `Xamarin.Android.Support.ViewPager`, or remove the reference to `Com.ViewPagerIndicator`.

**Comment by rysbilinski** [comment]
Hi @alexrainman is there any chance this will be fixed soon?

**Comment by nrb1238** [comment]
@rysbilinski , @PWaliaDev 
Better to try this https://github.com/AndreiMisiukevich/CardView

**Comment by rysbilinski** [comment]
> @rysbilinski , @PWaliaDev
> Better to try this https://github.com/AndreiMisiukevich/CardView

Thanks @nrb1238 I did this earlier today and working, but was just hoping to not have to make the change

**Comment by PWaliaDev** [comment]
@nrb1238, thanks for suggesting, actually this package does not have Scrolling event so thats why i can't go fo that.

**Comment by LeoJHarris** [comment]
@rysbilinski  @PWaliaDev @nrb1238 hang in there. There appears to be a proposal on xamarin forms team for carousel integrated in XF https://github.com/xamarin/Xamarin.Forms/issues/4996 watch that space

**Comment by NiknakSi** [comment]
I did suggest a temporary fix for this back in February, it solved the issue for me:

https://github.com/alexrainman/CarouselView/issues/489#issuecomment-465651202

All you need to do is clone the repo, compile, and use that instead of the Nuget.

**Comment by LeoJHarris** [comment]
@NiknakSi cloned your repo, complie:

throwing this when navigating away from the page with carousel:

System.NotSupportedException: Unable to activate instance of type Com.ViewPagerIndicator.CirclePageIndicator from native handle 0xbef9041c (key_handle 0x3d40df99).

**Comment by NiknakSi** [comment]
@LeoJHarris That's odd - did you remove any existing references for CirclePageIndicator and CarouselView from your projects? Do you have a sample project you could share?

**Comment by LeoJHarris** [comment]
@NiknakSi I simply used your cloned repo https://github.com/NiknakSi/CarouselView/tree/niknaksi/CirclePageIndicator-temporaryfix

But for now I have rolled back on Xamarin.Forms as well as this library and everything is working so will stick to that for now. 

But here's hoping a new release will be made on here soon. @alexrainman we would really appreciate some update on whether v6.0 be released soon or not?

**Comment by alexrainman** [comment]
@NiknakSi the answer is yes. I am into CarouselView development and i expect to release new version in couple of weeks.

**Comment by LeoJHarris** [comment]
@alexrainman Great! We will continue to look forward to release 6.0.  👍

**Comment by Abdulbasha** [comment]
@alexrainman After Updating XF4.0.0.425677 and Changed the Target Framework ( Android 8.1 ), Still i am facing the same issue and getting exception like Unable to activate instance of type Com.ViewPagerIndicator.CirclePageIndicator from native handle 0x7fe4ce06c4 . Can you please provide us the solution.

**Comment by BryanSan** [comment]
I had this same error and I was able to fix it installing the other package of alexrainman Xamarin.Android.CirclePageIndicator 1.0.3 and then it started to work.

I am using XF4.0.0.425677 with Target Framework Android 9

**Comment by alexrainman** [comment]
I am back! Releasing new version in the next 2 weeks. This issue will be corrected.

**Comment by Abdulbasha** [comment]
My issue is solved now. I Changed the Target Framework ( Android 9.0 ) and updated the Xamarin.Android.CirclePageIndicator to 1.0.3. Thanks to @BryanSan. And thanks to @alexrainman  for new release.

**Comment by Phenek** [comment]
I updated CirclePageIndicator to 1.0.3, 

Even if I install ViewPager I got the:
````
Can not resolve reference: Xamarin.Android.Support.ViewPager, referenced by Com.ViewPagerIndicator.
````

@BryanSan Do you have a workaround until two weeks?

**Comment by BryanSan** [comment]
@Phenek Try without ViewPager installed. I don't have it installed. Also, make sure you are targeting Android 9.0 and using the last Xamarin.Forms that is how I made it work

**Comment by Phenek** [comment]
No that didn't work
I Still got "Can not resolve reference: Xamarin.Android.Support.ViewPager"

**Comment by tunbeo** [comment]
It works for me,

I updated All Project to Xamarin Forms 4.0
In Android, add Xamarin.Android.CirclePageIndicator 1.0.3 , used Android 9.0, in Android Build checked Enable Multi-Dex. Then delete all bin and obj folders, open Solution again and it works like a charm hehe

**Comment by LeoJHarris** [comment]
@tunbeo what version this plugin you use? latest 5.0.1 or 5.0.2?

**Comment by LeoJHarris** [comment]
@tunbeo Followed as you mentioned but on returning to the page with the Carousel it didnt show any content inside it.

**Comment by LeoJHarris** [comment]
> Releasing new version in the next 2 weeks

@alexrainman Just a gentle reminder that we are still waiting on new release 6.0 and we need a resolution to this issue as soon as possible.

**Comment by Phenek** [comment]
@alexrainman This issue should be open, it has never been fixed.

If you create a branch with your 6.0 Preview, we could help.
because without we just see a repository never updated since 2 years.

#525 #508 #518

**Comment by picko94** [comment]
@alexrainman Should this issue be closed?

There hasn't been a new release and none of the solutions suggested in the above comments have resolved this issue for me.

**Comment by Phenek** [comment]
@picko94 you right!
the code isn't been updated since two years.
No PR accepted, no active working branch.

@alexrainman do some communication some time but we don't understand what is really happening to him. Never mind, I hope he's fine.

Xamarin Team will release[ Xam 4.3](https://github.com/xamarin/Xamarin.Forms/projects/39) in the next sprint with a Carousel implementation with bugs for sure...

**Comment by LeoJHarris** [comment]
Try https://github.com/AndreiMisiukevich/CardView it has carouselview and good support by author. We switched and very happy with it.

**Comment by lewcianci** [comment]
This package not working was one of the reasons I switched to flutter. They do it natively in their framework.

Video here: 
https://youtu.be/J1gE9xvph-A

I'm thinking about writing some articles on medium on how I switched from xamarin to flutter. Stuff like this is just a piece of cake now. If anyone would be interested in articles like that let me know and I will write some 😀

**Comment by alexrainman** [comment]
It has taken me more than expected to correct all the issues and add all the new functionality i wanted to add but the carousel is alive and will be out soon.

@Phenek i am just fine 😁

**Comment by srivastava1july** [comment]
When can we expect the updated nuget package for this issue. Its been long we are struggling with this issue.

**Comment by Ruddy2007** [comment]
> When can we expect the updated nuget package for this issue. Its been long we are struggling with this issue.

Don't hold your breath, we have been waiting for a reason for months. Even if it does get released it sounds to be a **huge** update. That will lead to a huge amount of bugs in one hit also, so this may get fixed but no doubt other bugs will appear (if we go by the amount of time the owner has closed bugs and said its fixed and will be in the next release). 

You would be better forking and fixing the issue yourself, that's what I have had to do.

**Comment by LeoJHarris** [comment]
@srivastava1july We are using https://github.com/AndreiMisiukevich/CardView or you use the provided carouselview in xamarin forms itself. Done.

**Comment by srivastava1july** [comment]
@Ruddy2007 How did you fix this issue at you end?

**Comment by srivastava1july** [comment]
@LeoJHarris The app is already live and we have created custom renderer as per our requirement. Using a new control at this stage will lead to lot of rework. That's why I want to fix this with current implementation.

**Comment by LeoJHarris** [comment]
@srivastava1july I think you might not have a choice but to bite the bullet and migrate as you might never see a new release on this project.

Personally I prefer to have depedancy on the Xamarin Forms framework where I can then one person open source project and progressively adding more dependancy to it, best to future proof your app developement.
